File: web/src/domain/time.ts
import type { IsoDate, Time24 } from "./primitives";

/**
 * Everything here works in the device's local calendar on purpose. A school day is a local
 * concept: converting to UTC would push an evening record into the next day for anyone east of
 * Greenwich, and silently corrupt every report built on top of it.
 */

const MINUTES_PER_DAY = 24 * 60;

export function toIsoDate(date: Date): IsoDate {
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, "0");
  const day = String(date.getDate()).padStart(2, "0");
  return `${year}-${month}-${day}`;
}

export function fromIsoDate(date: IsoDate): Date {
  const [year, month, day] = date.split("-").map(Number);
  return new Date(year, month - 1, day);
}

export function today(now: Date = new Date()): IsoDate {
  return toIsoDate(now);
}

/** Minutes elapsed since local midnight — the unit the window state machine compares in. */
export function minutesOfDay(now: Date = new Date()): number {
  return now.getHours() * 60 + now.getMinutes();
}

export function timeToMinutes(time: Time24): number {
  const [hours, minutes] = time.split(":").map(Number);
  return hours * 60 + minutes;
}

export function minutesToTime(minutes: number): Time24 {
  const wrapped = ((minutes % MINUTES_PER_DAY) + MINUTES_PER_DAY) % MINUTES_PER_DAY;
  const hours = String(Math.floor(wrapped / 60)).padStart(2, "0");
  const mins = String(wrapped % 60).padStart(2, "0");
  return `${hours}:${mins}`;
}

/** `14:05` renders as `2:05 PM` — the format the original app and its users expect. */
export function formatTime12(time: Time24): string {
  const [hours, minutes] = time.split(":").map(Number);
  const period = hours < 12 ? "AM" : "PM";
  const hour12 = hours % 12 === 0 ? 12 : hours % 12;
  return `${hour12}:${String(minutes).padStart(2, "0")} ${period}`;
}

/**
 * Parses the legacy Android format `h:mm:AM` (for example `7:30:AM`) into `HH:mm`.
 * Used only by the legacy-import adapter. Returns null on anything it does not recognise.
 */
export function parseLegacyTime(value: string): Time24 | null {
  const match = /^(\d{1,2}):([0-5]\d):(AM|PM)$/i.exec(value.trim());
  if (!match) return null;

  const [, rawHour, minutes, rawPeriod] = match;
  const hour = Number(rawHour);
  if (hour < 1 || hour > 12) return null;

  const period = rawPeriod.toUpperCase();
  const hours24 = period === "AM" ? hour % 12 : (hour % 12) + 12;
  return `${String(hours24).padStart(2, "0")}:${minutes}`;
}

export function formatDateLong(date: IsoDate): string {
  return fromIsoDate(date).toLocaleDateString(undefined, {
    weekday: "short",
    year: "numeric",
    month: "short",
    day: "numeric",
  });
}

export function formatDateShort(date: IsoDate): string {
  return fromIsoDate(date).toLocaleDateString(undefined, {
    month: "short",
    day: "numeric",
  });
}

export function addDays(date: IsoDate, days: number): IsoDate {
  const shifted = fromIsoDate(date);
  shifted.setDate(shifted.getDate() + days);
  return toIsoDate(shifted);
}

/** Inclusive on both ends. Returns [] if the range is inverted. */
export function datesBetween(start: IsoDate, end: IsoDate): IsoDate[] {
  if (start > end) return [];
  const dates: IsoDate[] = [];
  for (let cursor = start; cursor <= end; cursor = addDays(cursor, 1)) {
    dates.push(cursor);
  }
  return dates;
}

export function isWithinRange(date: IsoDate, start: IsoDate, end: IsoDate): boolean {
  return date >= start && date <= end;
}

/** First and last day of the month containing `date`. */
export function monthRange(date: IsoDate): { start: IsoDate; end: IsoDate } {
  const [year, month] = date.split("-").map(Number);
  const start = toIsoDate(new Date(year, month - 1, 1));
  const end = toIsoDate(new Date(year, month, 0));
  return { start, end };
}

File: mobile-android/app/src/main/java/ph/attendance/domain/Time.kt
package ph.attendance.domain

import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException

/**
 * Everything here works in the device's local calendar on purpose.
 *
 * A school day is a local concept: converting to UTC would push an evening record into the next
 * day for anyone east of Greenwich and silently corrupt every report built on it. Timestamps sent
 * over the wire are instants in UTC; dates are not.
 */
object Clocks {
    fun nowIso(instant: Instant = Instant.now()): String =
        DateTimeFormatter.ISO_INSTANT.format(instant.truncatedToMillis())

    private fun Instant.truncatedToMillis(): Instant =
        Instant.ofEpochMilli(toEpochMilli())

    fun today(zone: ZoneId = ZoneId.systemDefault()): String = LocalDate.now(zone).toString()

    fun minutesOfDay(zone: ZoneId = ZoneId.systemDefault()): Int =
        LocalTime.now(zone).let { it.hour * 60 + it.minute }

    fun minutesOf(dateTime: LocalDateTime): Int = dateTime.hour * 60 + dateTime.minute

    fun isoDate(dateTime: LocalDateTime): String = dateTime.toLocalDate().toString()

    /** An ISO instant for a local wall-clock moment. Used by the seed and by tests. */
    fun isoInstant(dateTime: LocalDateTime, zone: ZoneId = ZoneId.systemDefault()): String =
        DateTimeFormatter.ISO_INSTANT.format(dateTime.atZone(zone).toInstant())
}

fun timeToMinutes(time: String): Int {
    val parts = time.split(":")
    return parts[0].toInt() * 60 + parts[1].toInt()
}

fun minutesToTime(minutes: Int): String {
    val wrapped = ((minutes % 1440) + 1440) % 1440
    return "%02d:%02d".format(wrapped / 60, wrapped % 60)
}

/** `14:05` renders as `2:05 PM` — the format the original app and its users expect. */
fun formatTime12(time: String): String {
    val (hours, minutes) = time.split(":").map(String::toInt)
    val period = if (hours < 12) "AM" else "PM"
    val hour12 = if (hours % 12 == 0) 12 else hours % 12
    return "%d:%02d %s".format(hour12, minutes, period)
}

fun isValidTime24(value: String): Boolean =
    Regex("^([01]\\d|2[0-3]):[0-5]\\d$").matches(value)

fun isValidIsoDate(value: String): Boolean =
    try {
        LocalDate.parse(value)
        Regex("^\\d{4}-\\d{2}-\\d{2}$").matches(value)
    } catch (_: DateTimeParseException) {
        false
    }

fun addDays(date: String, days: Long): String = LocalDate.parse(date).plusDays(days).toString()

/** Inclusive on both ends; empty when the range is inverted. */
fun datesBetween(start: String, end: String): List<String> {
    if (start > end) return emptyList()
    val from = LocalDate.parse(start)
    val to = LocalDate.parse(end)
    return generateSequence(from) { current -> current.plusDays(1).takeIf { !it.isAfter(to) } }
        .map(LocalDate::toString)
        .toList()
}

fun monthRange(date: String): Pair<String, String> {
    val day = LocalDate.parse(date)
    return day.withDayOfMonth(1).toString() to day.withDayOfMonth(day.lengthOfMonth()).toString()
}

fun formatDateLong(date: String): String =
    LocalDate.parse(date).format(DateTimeFormatter.ofPattern("EEE, d MMM yyyy"))

fun formatDateShort(date: String): String =
    LocalDate.parse(date).format(DateTimeFormatter.ofPattern("d MMM"))

/**
 * Parses the legacy Android format `h:mm:AM` into `HH:mm`. Used only when importing data from
 * the 2019 app; returns null on anything it does not recognise.
 */
fun parseLegacyTime(value: String): String? {
    val match = Regex("^(\\d{1,2}):([0-5]\\d):(AM|PM)$", RegexOption.IGNORE_CASE)
        .matchEntire(value.trim()) ?: return null

    val hour = match.groupValues[1].toInt()
    if (hour !in 1..12) return null

    val minutes = match.groupValues[2]
    val isAm = match.groupValues[3].equals("AM", ignoreCase = true)
    val hours24 = if (isAm) hour % 12 else (hour % 12) + 12
    return "%02d:%s".format(hours24, minutes)
}
